### ISSUE TITLE
Serialization of Chart as dict not creating equivalent object

### ISSUE THREAD

**Issue by bergtholdt** [altair-viz/altair#1644]
When serializing a Chart as dictionary (and subsequently as json-string) and recreating a Chart from the dictionary, the created object is not equivalent to the original one based on the ``__eq__`` operator. Consider this example:

```python
import altair
from vega_datasets import data

iris = data.iris()
ch = altair.Chart(iris).mark_point().encode(
    x='petalLength',
    y='petalWidth',
    color='species'
)

ch1 = altair.Chart(iris).mark_point().encode(
    x='petalLength',
    y='petalWidth',
    color='species'
)
# works: same args and kwargs
assert ch == ch1

# "serialization" via dict
ch2 = altair.Chart.from_dict(ch.to_dict())
# AssertionError
assert ch == ch2
# works, the resulting dictionary representations are the same
assert ch.to_dict() == ch2.to_dict()
```

The problem is the ``__eq__`` operator in altari.utils.schemapi.SchemaBase:

```python
    def __eq__(self, other):
        return (type(self) is type(other)
                and self._args == other._args
                and self._kwds == other._kwds)
```

Changing it to 
```python
    def __eq__(self, other):
        return (type(self) is type(other)
                and self.to_dict() == other.to_dict())
```

Would work also for re-instantiated Charts from a dictionary. I created a pull request #1643 for the change.

**Comment by jakevdp** [comment]
This is by design – there is post-processing involved in serialization that changes the chart (e.g. applying themes, transforming and converting data, inferring data types, etc.)

**Comment by bergtholdt** [comment]
Resolved using pickle with #1706.
